lib/pbench/server/api/auth.py

```diff
@@ -0,0 +1,129 @@
+from pbench.server.database.models.active_tokens import ActiveTokens
+
+
+class Auth:
+    token_auth = HTTPTokenAuth("Bearer")
+
+    @staticmethod
+    def set_logger(logger):
+        # Logger gets set at the time of auth module initialization
+        Auth.logger = logger
+
+    def encode_auth_token(self, token_expire_duration, user_id):
+        """
+        Generates the Auth Token
+        :return: jwt token string
+        """
+        current_utc = datetime.datetime.utcnow()
+        payload = {
+            "iat": current_utc,
+            "exp": current_utc + datetime.timedelta(minutes=int(token_expire_duration)),
+            "sub": user_id,
+        }
+
+        # Get jwt key
+        jwt_key = self.get_secret_key()
+        return jwt.encode(payload, jwt_key, algorithm="HS256")
+
+    def get_secret_key(self):
+        try:
+            return os.getenv("SECRET_KEY", "my_precious")
+        except Exception as e:
+            Auth.logger.exception(f"{__name__}: ERROR: {e.__traceback__}")
+
+    def verify_user(self, username):
+        """
+        Check if the provided username belong to the current user by
+        querying the Usermodel with the current user
+        :param username:
+        :param logger
+        :return: User (UserModel instance), verified status (boolean)
+        """
+        user = User.query(id=self.token_auth.current_user().id)
+        # check if the current username matches with the one provided
+        verified = user is not None and user.username == username
+        Auth.logger.warning("verified status of user '{}' is '{}'", username, verified)
+
+        return user, verified
+
+    def get_auth_token(self, logger):
+        # get auth token
+        auth_header = request.headers.get("Authorization")
+
+        if not auth_header:
+            logger.warning("Missing expected Authorization header")
+            abort(
+                403,
+                message="Please add 'Authorization' token as Authorization: Bearer <JWT_Auth_token>",
+            )
+
+        try:
+            auth_schema, auth_token = auth_header.split()
+        except ValueError:
+            logger.warning("Malformed Auth header during logout")
+            abort(
+                401,
+                message="Malformed Authorization header, please add request header as Authorization: Bearer <session_token>",
+            )
+        else:
+            if auth_schema.lower() != "bearer":
+                logger.warning(
+                    "Expected authorization schema to be 'bearer', not '{}'",
+                    auth_schema,
+                )
+                abort(
+                    401,
+                    message="Malformed Authorization header, please add request header as Authorization: Bearer <session_token>",
+                )
+            return auth_token
+
+    def verify_auth(self):
+        @Auth.token_auth.verify_token
+        def verify_token(auth_token):
```

Comment:
okay so it works the following way:
1. If you look inside the `lib/pbench/server/api/resources/users_api.py` file, you will notice all our apis are protected with a decorator `@Auth.token_auth.login_required(f=Auth().verify_auth())`
2. It is defined with an `Auth` class member `token_auth` which is of type `HTTPTokenAuth("Bearer")` line 11 in `lib/pbench/server/api/auth.py`
3. Whenever a `login_required` is called on the `token_auth` it will go to a function defined by the parametr `f`. This is what happens at the beginning of our each rest API method 
4. According to `flask_httpauth` package the function to validate login_required need to be defined with decorator `@token_auth.verify_token`
5. So when inside the `verify_auth` method it looks for the function define with `verify_token` decorator and it takes a single argument (token) and return either false (in which case we would not proceed with the rest request) or a valid `user` (in which case we will proceed with the rest request)

Ref: [flask_httpauth](https://flask-httpauth.readthedocs.io/en/latest/)
The Auth class definitely need some refactoring here but I will address that in a separate PR. However, Does this makes sense?

Reply:
> all our apis are protected with a decorator `@Auth.token_auth.login_required(f=Auth().verify_auth())`

My concern is that `Auth().verify_auth()` evaluates to `None` (I think...), thus the decorator devolves to

```
@Auth.token_auth.login_required(f=None)
```
and I have _no idea_ what happens after that point....

Reply:
I see the confusion, I have updated the function so its more cleaner

Reply:
I commented on the updated function, so I'm resolving this conversation in favor of that one....